Sources/OpenAPIKit/Document/Document.swift

```diff
@@ -94,8 +94,16 @@ extension OpenAPI {
         /// share definitions than to just use the same Swift value multiple times, but
         /// you still might want to consider using the Components Object for its impact
         /// on the JSON/YAML structure of your document once encoded.
+      
         public var components: Components
-
+      
+        /// The incoming webhooks that MAY be received as part of this API and that the API consumer MAY choose to implement.
+        ///
+        /// Closely related to the callbacks feature, this section describes requests initiated other than by an API call, for example by an out of band registration.
+        /// The key name is a unique string to refer to each webhook, while the (optionally referenced) Path Item Object describes a request that may be initiated by the API provider and the expected responses
+      
+        public var webhooks: ComponentDictionary<PathItem>
```

Comment:
I have changed `webhooks` to use `OrderedStringDictionary` which is just an `OrderedDictionary<String, T>`. Will rename if this is not appropriate.

Reply:
@mattpolzin  I need a little assistance with:
"make this support JSON references to path items in the components object"

Thanks

Reply:
> Will rename if this is not appropriate.

I do think in this case we should probably just stick with directly using the `OrderedDictionary` - my thinking is that `OrderedStringDictionary<T>` does not make `OrderedDictionary<String, T>` shorter or easier to understand but it still adds an extra type for those of us reading the code to need to follow. I actually feel a bit unsure of the advantage of the similar `ComponentDictionary<T>` type, although that type is used a bunch of lines in a row so the noise reduction over `OrderedDictionary<ComponentKey, T>` _might_ add up enough to make it worthwhile in that case.

Either way, the component dictionary type will need to  be rethought by the end of the v3.0 push in order to support JSON references within the components object (e.g. from one component to another).

Reply:
>  I need a little assistance with:
> "make this support JSON references to path items in the components object"

Certainly, happy to say more about that.

There are a couple of small changes needed to support JSON references here.
1. We need to use an `Either`, similar to how [this property](https://github.com/mattpolzin/OpenAPIKit/blob/release/3_0/Sources/OpenAPIKit/Operation/Operation.swift#L30) supports JSON references to operations _or_ directly inserting the Operation object.
2. The `Either+Convenience.swift` file should get a getter extension [like this](https://github.com/mattpolzin/OpenAPIKit/blob/release/3_0/Sources/OpenAPIKit/Either/Either%2BConvenience.swift#L76) and a constructor extension [like this](https://github.com/mattpolzin/OpenAPIKit/blob/release/3_0/Sources/OpenAPIKit/Either/Either%2BConvenience.swift#L167) for when `B` is `OpenAPI.PathItem`.

Then there's probably a good tweak to be made to the `init` for `OpenAPI.Document` to allow the user to specify either a JSON reference or a path item.

Reply:
@mattpolzin I have updated the PR with your suggestions. Thanks.
I tried to add the support to JSON references like this:

`public var pathItems: Either<JSONReference<ComponentDictionary<PathItem>>, ComponentDictionary<PathItem>>`

That produced an error that makes no sense :

JSONReference' requires the types 'OpenAPI.ComponentKey' and 'OpenAPI.CallbackURL' be equivalent

Reply:
That is certainly not a very clear error. 

I do see a couple of issues, but I'm not immediately sure how to tie the error message to either of them!

The `Either` type you have is sensible based on the reference type you wrote but not quite what we need in this case because we want "a dictionary with string keys and either path item or reference values." The `JSONReference` type expects as its specialization parameter the thing we are going to be able to look up in the components dictionary, in this case a path item.

Based on those two things, the type we need for `webhooks` is
```Swift
OrderedDictionary<String, Either<JSONReference<OpenAPI.PathItem>, OpenAPI.PathItem>>
```

I'm not at the computer so I may not have gotten that perfectly accurate but hopefully it's a helpful hint at least. I can confirm I've written a valid type signature later today!

Reply:
Thanks for the explanation.

Reply:
@mattt I have updated the PR with your suggestions and merged it with the base `3_0` branch.

I've also investigated further all types that use `Either` in an attempt to discover a pattern for adding such type definitions.
I've made a gist: https://gist.github.com/mihaelamj/ebd4e0cd7b905d67aef7ad31c88ad5e2 
demonstrating my attempt to unify all types by extracting common elements.
For example, `webhooks` are just a typename for a map of `Either<JSONReference<Type>, Type>` items.
As such, it should be defined somewhere in the `OpenAPI.PathItem` namespace scope, and used from different elements, if needed.

Reply:
This is some really great analysis. Exactly the kind of process I go through too infrequently in the course of developing a project. 

A note about the following comment: I've really let myself go on, so you might be getting more of a  response than you signed up for!

First, I'll say up front, the type of the new `webhooks` property is definitely a mouthful, no doubt about it: `OrderedDictionary<String, Either<JSONReference<OpenAPI.PathItem>, OpenAPI.PathItem>>`. Also, you've absolutely picked one of the most common recurring patterns (`Either` uses) to investigate, and I think you've done a great job of boiling down the various situations where the type comes up.

Now to talk a bit more philosophically about this (which is to say: opinions vary, these are mine, and I've seen my taste change even since I started writing OpenAPIKit).

As time goes on, I find myself using type aliases less rather than more. I find myself wanting to really motivate the use of type aliases by (a) a new name being drastically different in meaning than the name being typealiased, (b) Part of a generic type signature being cleanly renamed to remove boilerplate without adding mental overhead, (c) relating a generic type to a specific use-case motivated by factors outside the project.

These all deserve examples before I go on.

An example of (a) might be that I am coding something related to email addresses, and I have no immediate intention of validating that an email is actually an email; I might find it worth creating a typealias for `String` named `Email`. It's clear what I expect anywhere I require that type, and I can even swap String out for a type that does offer additional validation of emails later without changing as much code.

An example of (b) might  be that I have a generic `Gradient<ColorA, ColorB>` type that represents a color gradient and I create a `Solid<Color>` typealias for `Gradient<Color, Color>`; The semantics of the name I chose for the typealias fully explain why only 1 `Color` specialization is needed so I've simplified a type signature without losing any clarity.

An example of (c) lands really close to home in this project, where I chose, for example, to typealias `OrderedDictionary<OpenAPI.Path, OpenAPI.PathItem>` as `OpenAPI.PathItem.Map`; I did not do this because my typealias name is shorter or clearer but rather because the OpenAPI specification defines a ["Paths Object"](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.1.0.md#paths-object) for which I wanted a name in the codebase. Having the name `OpenAPI.PathItem.Map` allows me to more clearly document the translation in the [Project Status](https://github.com/mattpolzin/OpenAPIKit/blob/main/documentation/specification_coverage.md#paths-object-openapipathitemmap). It's a matter of opinion whether that offers clarity, but these are after all _my thoughts_ on the subject.

Those might be my motivations for creating typealiases, but then there's my proclivity for how to structure them. I think the simplest most powerful motivator for me there is I try not to hide the structure of my types in typealises. For example, I will prefer a `Map<Person>` type that I can read as "A map of people" over a `Person_Map` type that I can surmise is a "map" of some kind by its name, but it is less clear whether it is structurally a map type with a person within it and definitely unclear if this is the same "map" as the one referred to by `Animal_map` (whereas I can make pretty strong assumptions that `Map<Person>` and `Map<Animal>` share a common structure).

So now I can talk more about the `Either` examples you analyzed in that "framework" (while fully admitting that there are some typealiases I've written, even in this project, that I probably would not find worthwhile if I revisited the decision today).

You've suggested `RefOrSelf` namespaced by various resources types in the codebase. This actually is rather nice, in my opinion. The type of reference is given by the namespace and a lot of boilerplate goes away without losing clarity: Of course a `JSONSchema.RefOrSelf` is "either a reference to a JSON Schema or a JSON Schema" (i.e. `Either<JSONReference<JSONSchema>, JSONSchema>`).

The trickier sell in this case is the name in particular. As you noted, it's a bit abrupt (to use my own words to describe it); also importantly, it does not follow the SSWG guidelines to use complete words in names whenever possible instead of abbreviating them. I've had similar thoughts to yours before, but I was unsatisfied with the names I came up with as well: `MaybeReference` or `Referenceable<T>`. If you spell your suggestion out fully, `ReferenceOrSelf` I like it more than my names, although "reference" and "self" are pretty overloaded terms unfortunately so perhaps there is a _bit_ of clarity lost of the full type which specifies it is talking about a `JSONReference` in particular. I'm not totally opposed to this type alias when spelled out fully or perhaps with a subtle name tweak, honestly.

Because the "or" in `RefOrSelf` plays the same _role_ as `Either` does (it lets me know there are two alternatives, I also start to wonder if it would play out nicely if you left the `Either` in place and just typealiased what it meant to be a particular kind of reference:

```swift
extension JSONSchema {
    public typealias Reference = JSONReference<JSONSchema>
}
...
let schema: Either<JSONSchema.Reference, JSONSchema>
```

With this breakdown, I would argue you don't gain very much, but there is a bit less visual clutter than with `Either<JSONReference<JSONSchema>, JSONSchema>`.

The typealiases you proposed that have `_Map` or `_Array` suffixes would be a bit harder for me to justify under the framework I made up. 

The `_Map` suffix consistently takes the place of `OrderedDictionary<String, T>`, which loses a bit of clarity because there's no particular reason why a "Map" should have `String` keys, it just happens to be the case that it does here. That's maybe not so big of a deal, but you might in this case get more clarity out of a common alias like `OpenAPI.Map<T>` aliasing `OrderedDictionary<String, T>` such that you get types like `OpenAPI.Map<OpenAPI.Example.RefOrSelf>` instead of `OpenAPI.Example.RefOrSelf_Map`. The length is slightly longer but I give it more points for clarity because I'm used to reading types as "this is a map _of_ that" and I don't need to wonder what secrets lie beneath the map-like thing called `OpenAPI.Example.RefOrSelf_Map`.

In the latter case, you trade something like `[JSONSchema.RefOrSelf]` for something like `RefOrSelf_Array`; the former is very recognizable Swift code whereas the latter leaves me wondering if there is something really special about this particular "array" such that it is not really an `Array` a I know them.

----------

At the end of all of this, I come away without feeling like I absolutely _must_ have my hands on shorter type aliases for `Either<JSONReference<T>, T>` _but_ I think you've made a compelling case, some great observations, and some good suggestions.

Finally, to revisit the monster that is `OrderedDictionary<String, Either<JSONReference<OpenAPI.PathItem>, OpenAPI.PathItem>>` -- that is just too darn long to ignore; I _do_ feel compelled to make the line using that type a bit easier to digest when I read it to myself!

It's interesting, then to compare and contrast a few alternatives from your post and mine:
```swift
// as-is
OrderedDictionary<String, Either<JSONReference<OpenAPI.PathItem>, OpenAPI.PathItem>>

// as-proposed
Webhooks
// - or -
OpenAPI.Pathitem.RefOrSelf_Map

// another possibility
OpenAPI.Map<Referenceable<OpenAPI.PathItem>>
```

Of them all, I think I am drawn to the way that the last option hides as little about the structure as it needs to: it is unapologetically a "map of path items that could be references." Still, I simultaneously shy away from the fact that it seemingly arbitrarily refers to a `String`-keyed map 🤷.

Reply:
Thank you for a comprehensive reply. It was an interesting discussion.